Subject: [PATCH] change return value
---
Index: src/main/kotlin/com/deshyt/buffered/BufferedChannel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/kotlin/com/deshyt/buffered/BufferedChannel.kt b/src/main/kotlin/com/deshyt/buffered/BufferedChannel.kt
--- a/src/main/kotlin/com/deshyt/buffered/BufferedChannel.kt	(revision 221dbb474de5cb1ad6b006ce19a59ffc8787bcb4)
+++ b/src/main/kotlin/com/deshyt/buffered/BufferedChannel.kt	(revision bb90a19f5a4e82421731a76b0c00d29ef9ab26b6)
@@ -71,8 +71,20 @@
             segment.setElement(index, elem)
             // Update the cell according to the algorithm. If the cell was poisoned or
             // stores an interrupted receiver, clean the cell and restart the sender.
-            if (updateCellOnSend(s, segment, index)) return
-            segment.cleanElement(index)
+            when (updateCellOnSend(s, segment, index)) {
+                // The element was buffered => finish
+                RESULT_BUFFERED -> return
+                // The rendezvous happened => finish
+                RESULT_RENDEZVOUS -> return
+                // The sender was suspended => finish, the opposite request was responsible for the progress
+                RESULT_SUSPENDED -> return
+                // The cell was poisoned or stores an interrupted receiver => restart
+                RESULT_FAILED -> {
+                    // Clean the current cell and restart.
+                    segment.cleanElement(index)
+                    continue
+                }
+            }
         }
     }
 
@@ -82,7 +94,7 @@
         /* The working cell is specified by the segment and the index in it. */
         segment: ChannelSegment<E>,
         index: Int
-    ): Boolean {
+    ): Int {
         while (true) {
             when (val state = segment.getState(index)) {
                 // The cell is empty.
@@ -93,33 +105,33 @@
                         // Move the cell state to `BUFFERED`.
                         if (segment.casState(index, null, CellState.BUFFERED)) {
                             // The element has been successfully buffered, finish.
-                            return true
+                            return RESULT_BUFFERED
                         }
                     } else {
                         // The sender should suspend.
-                        if (trySuspendRequest(segment, index, isSender = true)) return true
+                        if (trySuspendRequest(segment, index, isSender = true)) return RESULT_SUSPENDED
                     }
                 }
                 // The cell is in the logical buffer => try to buffer the element
                 CellState.IN_BUFFER -> {
-                    if (segment.casState(index, state, CellState.BUFFERED)) return true
+                    if (segment.casState(index, state, CellState.BUFFERED)) return RESULT_BUFFERED
                 }
                 // The cell was poisoned by a receiver => restart the sender
-                CellState.POISONED -> return false
+                CellState.POISONED -> return RESULT_FAILED
                 // Cancelled receiver => restart the sender
-                CellState.INTERRUPTED_RCV -> return false
+                CellState.INTERRUPTED_RCV -> return RESULT_FAILED
                 // Suspended receiver in the cell => try to resume it
                 else -> {
                     val receiver = (state as? Coroutine)?.cont ?: (state as CoroutineEB).cont
                     return if (tryResumeRequest(receiver)) {
                         segment.setState(index, CellState.DONE_RCV)
-                        true
+                        RESULT_RENDEZVOUS
                     } else {
                         // The resumption has failed, since the receiver was cancelled.
                         // Clean the cell and wait until `expandBuffer()`-s invoked on the cells
                         // before the current one finish.
                         segment.onCancellation(index = index, isSender = false)
-                        false
+                        RESULT_FAILED
                     }
                 }
             }
@@ -149,8 +161,13 @@
             // Update the cell according to the algorithm. If the rendezvous happened,
             // the received value is returned, then the cell is cleaned to avoid memory
             // leaks. Otherwise, the receiver restarts.
-            if (updateCellOnReceive(r, segment, index)) {
-                return segment.retrieveElement(index)
+            when (updateCellOnReceive(r, segment, index)) {
+                // The receiver was suspended => finish, the opposite request was responsible for the progress
+                RESULT_SUSPENDED -> return segment.retrieveElement(index)
+                // The rendezvous happened => finish
+                RESULT_RENDEZVOUS -> return segment.retrieveElement(index)
+                // The cell was poisoned or stores an interrupted sender => restart
+                RESULT_FAILED -> continue
             }
         }
     }
@@ -161,7 +178,7 @@
         /* The working cell is specified by the segment and the index in it. */
         segment: ChannelSegment<E>,
         index: Int
-    ): Boolean {
+    ): Int {
         while (true) {
             when (val state = segment.getState(index)) {
                 // The cell is empty.
@@ -177,20 +194,20 @@
                         if (segment.casState(index, state, CellState.POISONED)) {
                             // When the cell becomes poisoned, expand the logical buffer and restart.
                             expandBuffer()
-                            return false
+                            return RESULT_FAILED
                         }
                     } else {
                         // The receiver should suspend.
-                        if (trySuspendRequest(segment, index, isSender = false)) return true
+                        if (trySuspendRequest(segment, index, isSender = false)) return RESULT_SUSPENDED
                     }
                 }
                 // Buffered element => finish
                 CellState.BUFFERED -> {
                     segment.setState(index, CellState.DONE_RCV).also { expandBuffer() }
-                    return true
+                    return RESULT_RENDEZVOUS
                 }
                 // Cancelled sender => restart
-                CellState.INTERRUPTED_SEND -> return false
+                CellState.INTERRUPTED_SEND -> return RESULT_FAILED
                 // `expandBuffer()` is resuming the sender => wait
                 CellState.RESUMING_BY_EB -> continue
                 // Suspended sender in the cell => try to resume it
@@ -208,14 +225,14 @@
                             // In case a concurrent `expandBuffer()` has delegated its completion, the procedure should
                             // finish, as the sender is resumed. Thus, no further action is required.
                             segment.setState(index, CellState.DONE_RCV).also { expandBuffer() }
-                            true
+                            RESULT_RENDEZVOUS
                         } else {
                             // The resumption has failed. Update the cell state and restart the receiver.
                             // In case a concurrent `expandBuffer()` has delegated its completion, the procedure should
                             // skip this cell, so `expandBuffer()` should be called once again.
                             segment.onCancellation(index = index, isSender = true)
                             if (helpExpandBuffer) expandBuffer()
-                            false
+                            RESULT_FAILED
                         }
                     }
                 }
@@ -618,13 +621,21 @@

+/**
+   These constants are return values for [BufferedChannel.updateCellOnSend] and [BufferedChannel.updateCellOnReceive] methods.
+ */
+private const val RESULT_RENDEZVOUS = 0
+private const val RESULT_BUFFERED = 1
+private const val RESULT_SUSPENDED = 2
+private const val RESULT_FAILED = 3